source/CIS/cis-4.rst

```diff
@@ -0,0 +1,705 @@
+- The ``RevokeCredentialOtherParam``'s ``signing_data`` MUST include a nonce to protect against replay attacks.
+  The revocation authority's nonce is sequentially increased every time a revocation request is successfully executed.
+  The function MUST only accept a ``RevokeCredentialOtherParam`` if it has the next nonce following the sequential order.
+- The revocation MUST fail if:
+    - The credential ID is not present in the registry.
+    - The revocation key in unknown.
+    - The credential status is not one of ``Active`` or ``NotActivated`` (see :ref:`CIS-4-functions-credentialStatus`).
+    - The signature was intended for a different contract.
+    - The signature was intended for a different entrypoint.
+    - The signature is expired.
+    - The signature can not be validated.
+      The smart contract logic SHOULD practice its best efforts to ensure that only the revocation authority can generate and authorize a revocation request with a valid signature.
+
+.. _CIS-4-functions-registerRevocationKeys:
+
+``registerRevocationKeys``
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Register public keys that can be used by revocation authorities.
+
+Parameter
+~~~~~~~~~
+
+It is serialized as First 2 bytes encode the length (``n``) the vector of kesy, followed by this many :ref:`CIS-4-PublicKeyEd25519` keys::
+
+  RegisterPublicKeyParameters ::= (n: Byte²) (key: PublicKeyEd25519)ⁿ
+
+Requirements
+~~~~~~~~~~~~
+
+- The revocation MUST fail if:
+    - The the transaction ``sender`` is not the issuer.
+    - Some of the keys are already registered.
+- The smart contract MUST prevent resetting the nonce associated with a public key.
+  For example, the contract logic could keep track of all keys seen by the contract and avoid reusing the same keys even after the keys was made unavailable by calling :ref:`CIS-4-functions-removeRevocationKeys`.
+
+
+.. _CIS-4-functions-removeRevocationKeys:
+
+``removeRevocationKeys``
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Make a list of public keys unavailable to revocation authorities.
+
+Parameter
+~~~~~~~~~
+
+It is serialized as: First 2 bytes encode the length (``n``) the vector of keys, followed by this many :ref:`CIS-4-PublicKeyEd25519` keys::
+
+  RegisterPublicKeyParameters ::= (n: Byte²) (key: PublicKeyEd25519)ⁿ
+
+Requirements
+~~~~~~~~~~~~
+
+- The revocation MUST fail if:
+    - The the transaction ``sender`` is not the issuer.
+    - Some of the keys are unknown.
+
+
+.. _CIS-4-functions-revocationKeys:
+
+``revocationKeys``
+^^^^^^^^^^^^^^^^^^
+
+Query revocation keys.
+
+Response
+~~~~~~~~
+
+The function output a list of available revocation keys.
+Valid signatures with the corresponding private keys can be used to revoke any credential in the registry.
+
+It is serialized as: First 2 bytes encode the length (``n``) the vector of keys, followed by this many :ref:`CIS-4-PublicKeyEd25519` keys::
+
+  RegisterPublicKeyParameters ::= (n: Byte²) (key: PublicKeyEd25519)ⁿ
+
+
+Logged events
```

Comment:
Yes, this is a good point.
I think @abizjak was against it, but I don't remember why :)

Reply:
You mean not to make them clash with tags from other standards?

While it's true you cannot make CIS-2 + CIS-4 supporting contract, it is hard to imagine such a use case.

Having to have all distinct tags from different standards is hopeless.

I don't see this being a problem and I don't want to change things yet again.

Reply:
If this was a concern the only real solution would be to include the name of the standard in the event.

So we'd have events

```
struct Event {
   standard: Standard,
   standard_event: StandardDefinedEvent
}
```
where `Standard` would be names, so `CIS-0`, `CIS-1`, ...
and would be emitting these instead of requiring a globally unique u8 tag.

Reply:
Or instead of the full `CIS-?` it could just be a u16 or u32 with the standard ID.

Reply:
If you think there is a case where this will be a problem we can still change it now, it's just annoying since there are many components.

Reply:
There is also CIS3 which has this 250 event for nonce. It is conceivable to sponsor issuance perhaps so we need to then not use 250.

I don't want to spend more time revising serialization so we'll then just use 249.. codes for now.

But any future standard we must revise this, because it is not tenable to cram all possible event types into a u8 globally.

Reply:
I like the idea of using standard ids as tags in future CIS  standards.

Reply:
I fixed the event tags. Now they start with 249 counting down.

Reply:
yes, we should have a better solution in the future